p2p/src/sync/peer.rs

```diff
@@ -13,56 +13,157 @@
+// TODO: Track the block availability for a peer.
+// TODO: Track the best known block for a peer and take into account the chain work when syncing.
+/// A peer context.
+///
+/// Syncing logic runs in a separate task for each peer.
+pub struct Peer<T: NetworkingService> {
+    id: ConstValue<PeerId>,
+    p2p_config: Arc<P2pConfig>,
+    chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
+    peer_manager_sender: UnboundedSender<PeerManagerEvent<T>>,
+    message_sender: UnboundedSender<(PeerId, SyncMessage)>,
+    events_receiver: UnboundedReceiver<PeerEvent>,
+    is_initial_block_download: Arc<AtomicBool>,
+    /// A list of headers received via the `HeaderListResponse` message that we haven't yet
+    /// requested the blocks for.
+    known_headers: Vec<BlockHeader>,
+    /// A list of blocks that we requested from this peer.
+    requested_blocks: BTreeSet<Id<Block>>,
+    /// A queue of the blocks requested this peer.
+    blocks_queue: VecDeque<Id<Block>>,
+}
+
+impl<T> Peer<T>
 where
     T: NetworkingService,
     T::SyncingMessagingHandle: SyncingMessagingService<T>,
 {
-    pub(super) async fn handle_message(
-        &mut self,
-        peer: PeerId,
-        message: SyncMessage,
-    ) -> Result<()> {
-        match message {
-            SyncMessage::HeaderListRequest(r) => {
-                self.handle_header_request(peer, r.into_locator()).await
+    pub fn new(
+        id: PeerId,
+        p2p_config: Arc<P2pConfig>,
+        chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
+        peer_manager_sender: UnboundedSender<PeerManagerEvent<T>>,
+        message_sender: UnboundedSender<(PeerId, SyncMessage)>,
+        events_receiver: UnboundedReceiver<PeerEvent>,
+        is_initial_block_download: Arc<AtomicBool>,
+    ) -> Self {
+        Self {
+            id: id.into(),
+            p2p_config,
+            chainstate_handle,
+            peer_manager_sender,
+            message_sender,
+            events_receiver,
+            is_initial_block_download,
+            known_headers: Vec::new(),
+            requested_blocks: BTreeSet::new(),
+            blocks_queue: VecDeque::new(),
+        }
+    }
+
+    /// Returns an identifier of the peer associated with this task.
+    pub fn id(&self) -> PeerId {
+        *self.id
+    }
+
+    pub async fn run(&mut self) -> Result<Void> {
+        self.request_headers().await?;
+
+        loop {
+            tokio::select! {
+                event = self.events_receiver.recv() => {
+                    let event = event.ok_or(P2pError::ChannelClosed)?;
+                    self.handle_event(event).await?;
+                },
+
+                block = async { self.blocks_queue.pop_front().expect("The block queue is empty") }, if !self.blocks_queue.is_empty() => {
+                    let res = self.send_block(block).await;
+                    self.handle_result(res).await?;
```

Comment:
This is some weird logic I don't understand. Why are we handling both sending and receiving blocks to the peer with the same function?

When a peer sends us a block, that can be an attack, no question there.
But when we're sending a block, this can never be an attack.

These aren't the same.